### PR TITLE
Added a ClosureTask implementation for easier task scheduling

### PR DESCRIPTION
**Pull request by dktapps**
## Problem
It's very common for plugins to use scheduled tasks to fire functions within themselves. Before PocketMine-MP API 2.0.0 (henceforth referred to as 2.0 for simplicity), we had `CallbackTask` which allowed plugins to do this. However, in 2.0 the `CallbackTask` was removed.

While `CallbackTask` was never intended for plugin usage, the usefulness of callables in this aspect is undeniable. So, meet the new `ClosureTask`.

Consider the old workflow:

```php
//Main.php
$this->getScheduler()->scheduleDelayedTask(new SimpleTask($this), 1);

//SimpleTask.php
class SimpleTask{
    private $plugin;

    public function __construct(Main $plugin){
        $this->plugin = $plugin;
    }

    public function onRun(int $currentTick) : void{
        $this->plugin->getServer()->broadcastMessage("Hello players");
    }
}
```
This is very annoying especially when the `onRun()` body is merely 1 line. This can be reduced by using anonymous classes, but still causes a lot of boilerplate.

## Solution
Wouldn't it be nice to just be able to do this...
```php
//Main.php
$this->getScheduler()->scheduleDelayedTask(new ClosureTask(function(int $currentTick) : void{
    $this->getServer()->broadcastMessage("Hello players");
}, 1);
```

Closures are extremely powerful. The main reason is because they retain the `this` scope where they were created. They can also keep variables from the call site by `use`ing them, like `function() use($var1, $var2, $var3)`.

### API changes
- Added a new `ClosureTask` class. The constructor accepts a `Closure` which should accept exactly ONE parameter `(int $currentTick)` and SHOULD NOT return a value.

## Tests
```php
<?php

declare(strict_types=1);

require 'vendor/autoload.php';

$sched = new \pocketmine\scheduler\TaskScheduler(new \SimpleLogger());

$sched->scheduleTask(new \pocketmine\scheduler\ClosureTask(function(int $currentTick) : void{
	echo "hello sunshine!" . PHP_EOL;
	echo $currentTick . PHP_EOL;
}));

$sched->mainThreadHeartbeat(0);
```
outputs
```
hello sunshine!
0
```

### Unsolved issues
Timings task names are a problem which is currently unsolved, since they'll all appear as a generic `ClosureTask`.

### CODE REVIEW

File: src/pocketmine/scheduler/ClosureTask.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+namespace pocketmine\scheduler;
+
+/**
+ * Task implementation which allows closures to be called by a scheduler.
+ *
+ * Example usage:
+ *
+ * TaskScheduler->scheduleTask(new ClosureTask(function(int $currentTick) : void{
+ *     echo "HI on $currentTick\n";
+ * });
```
Comment (SOF3) [inline src/pocketmine/scheduler/ClosureTask.php]: This code snippet might not render properly in some documentation generators.

Comment (dktapps) [reply src/pocketmine/scheduler/ClosureTask.php]: I'm not sure how else to do it. Suggestions?

File: src/pocketmine/scheduler/ClosureTask.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+namespace pocketmine\scheduler;
+
+use pocketmine\utils\Utils;
+
+/**
+ * Task implementation which allows closures to be called by a scheduler.
+ *
+ * Example usage:
+ *
+ * TaskScheduler->scheduleTask(new ClosureTask(function(int $currentTick) : void{
```
Comment (SOF3) [inline src/pocketmine/scheduler/ClosureTask.php]:
```suggestion
 * ```
 * TaskScheduler->scheduleTask(new ClosureTask(function(int $currentTick) : void{
```

File: src/pocketmine/scheduler/ClosureTask.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ *     echo "HI on $currentTick\n";
+ * });
```
Comment (SOF3) [inline src/pocketmine/scheduler/ClosureTask.php]:
```suggestion
 * });
 * ```
```

**Review by SOF3 — APPROVED** [review]

**Comment by dktapps** [comment]
Considering extending this to allow any callable instead of just Closure objects. The implementation would be nearly identical to the current one.

**Pull request merged**